crates/pallets/starknet/src/transaction_validation.rs

```diff
@@ -50,3 +53,77 @@ impl<OuterOrigin: Into<Result<RawOrigin, OuterOrigin>> + From<RawOrigin>> Ensure
+                    Err(InvalidTransaction::Stale)?;
+                }
+
+                // A transaction with a nonce higher than the expected nonce is placed in
+                // the future queue of the transaction pool.
+                if sender_nonce < *transaction_nonce {
+                    log!(
+                        info,
+                        "Nonce is too high. Expected: {:?}, got: {:?}. This transaction will be placed in the \
+                         transaction pool and executed in the future when the nonce is reached.",
+                        sender_nonce,
+                        transaction_nonce
+                    );
+                }
+
+                Ok((tx.sender_address(), Some(sender_nonce), *transaction_nonce))
+            }
+            UserAndL1HandlerTransaction::L1Handler(tx, _fee) => {
+                Self::ensure_l1_message_not_executed(&Nonce(StarkFelt::from(tx.nonce)))?;
+                Ok((ContractAddress::default().into(), None, tx.nonce.into()))
+            }
+        }
+    }
+
+    pub fn validate_unsigned_tx(transaction: &UserAndL1HandlerTransaction) -> Result<(), InvalidTransaction> {
+        let chain_id = Self::chain_id();
+        let block_context = Self::get_block_context();
+        let mut state: BlockifierStateAdapter<T> = BlockifierStateAdapter::<T>::default();
+        let mut execution_resources = ExecutionResources::default();
+        let mut initial_gas = blockifier::abi::constants::INITIAL_GAS_COST;
+
+        match transaction {
+            UserAndL1HandlerTransaction::User(transaction) => {
+                match transaction {
+                    // There is no way to validate it before the account is actuallly deployed
+                    UserTransaction::DeployAccount(_) => Ok(None),
+                    UserTransaction::Declare(tx, contract_class) => tx
+                        .try_into_executable::<T::SystemHash>(chain_id, contract_class.clone(), false)
+                        .map_err(|_| InvalidTransaction::BadProof)?
+                        .validate_tx(&mut state, &block_context, &mut execution_resources, &mut initial_gas, false),
+                    UserTransaction::Invoke(tx) => tx.into_executable::<T::SystemHash>(chain_id, false).validate_tx(
+                        &mut state,
+                        &block_context,
+                        &mut execution_resources,
+                        &mut initial_gas,
+                        false,
+                    ),
+                }
+            }
+            UserAndL1HandlerTransaction::L1Handler(transaction, fee) => transaction
+                .into_executable::<T::SystemHash>(chain_id, *fee, false)
+                .validate_tx(&mut state, &block_context, &mut execution_resources, &mut initial_gas, false),
+        }
+        .map_err(|_| InvalidTransaction::BadProof)?;
+
+        Ok(())
+    }
+
+    pub fn ensure_l1_message_not_executed(nonce: &Nonce) -> Result<(), InvalidTransaction> {
```

Comment:
When this function is called, its error is either mapped to another error or converted to a bool. 

rename as `is_l1_message_already_executed` return a bool